### DISCUSSION TITLE
ctypes IntelliSense/Autocomplete

### DISCUSSION THREAD

**Discussion by TheRealQuantam** [microsoft/pylance-release#4438] [Enhancement]
**Describe the feature you are requesting**
Python has by far the best support for binary structured data of any scripting language I'm familiar with in ctypes - better even than many compiled languages. PTVS supports IS/AC for classes, named tuples, and probably others I'm not familiar with. However, it does not appear to support ctypes structures et al, instead only showing the members of the metaclass. This issue requests IS/AC support for ctypes structures/unions/etc. in at least Visual Studio 2022+.

**Comment by StellaHuang95** [comment]
Hi @TheRealQuantam, thanks for the suggestion. Could you please provide a code example that illustrates the problem you mentioned about PTVS not supporting ctypes structures and instead only displaying the members of the metaclass. By providing this code snippet, we will be able to have a clearer understanding of the specific feature you'd like to see in PTVS. Thank you!

**Comment by TheRealQuantam** [comment]
Oddly enough while getting a screen shot I got two very similar but slightly different behaviors:

![image](https://github.com/microsoft/PTVS/assets/54358053/311362d6-e46b-46b5-bcca-44bc325deee8)
![image](https://github.com/microsoft/PTVS/assets/54358053/1058ffc2-8ab9-4668-8413-19ddc1bb13d8)

```python
from ctypes import *

class NsfHeader(LittleEndianStructure):
	_pack_ = True
	_fields_ = (
		("magic", c_char * 5),
		("version", c_uint8),
		("num_tracks", c_uint8),
		("start_track", c_uint8), # 1-based
		("load_addr", c_uint16),
		("init_addr", c_uint16),
		("play_addr", c_uint16),
		("game_name", c_char * 32),
		("artist_name", c_char * 32),
		("copyright_name", c_char * 32),
		("ntsc_speed", c_uint16),
		("start_banks", c_uint8 * 8),
		("pal_speed", c_uint16),
		("machine_flags", c_uint8),
		("expansion_chips", c_uint8),
		("_reserved", c_uint8),
		("prg_size_lo", c_uint16),
		("prg_size_hi", c_uint8),
	)

hdr = NsfHeader()

hdr.
```

In either case, it's showing inherited members but not the fields of the structure itself (no, they're not further scrolled down in the list).

This is on:
Microsoft Visual Studio Community 2022
Version 17.6.2
VisualStudio.17.Release/17.6.2+33723.286
Python with Pylance   17.0.23087.1

**Comment by StellaHuang95** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by CaioMizerkowski** [comment]
To me this is an essencial feature.

**Comment by chrismeunier** [comment]
Any advancements on this feature ? This seems like a logical out-of-the-box feature to have.
In my case I am generating large ctypes structures (with sometimes nested structures) and it would be so much easier to work with those if the fields naturally showed up in the autocomplete popup.

**Reply by rchiodo** [reply]
It's unclear to me what the ask is. 

Are we talking about at the fields level you want the tuples there to be supported?

<img width="706" height="376" alt="image" src="https://github.com/user-attachments/assets/05f51a08-0fe9-4c05-a32d-a08c8c426722" />


Or should they be just at the `hdr.` level?

**Reply by rchiodo** [reply]
I asked copilot. It said the normal way would be to access them on the header structure itself. Like this:

```
hdr.start_banks
```

**Reply by chrismeunier** [reply]
> I asked copilot. It said the normal way would be to access them on the header structure itself. Like this:
> 
> ```
> hdr.start_banks
> ```

Yes exactly at the `hdr.` level, this is how `ctypes.Structure` fields are supposed to be accessed. The `_fields_` is only used for the initialization of the structure.

For instance here:
<img width="594" height="120" alt="image" src="https://github.com/user-attachments/assets/709f829a-d409-4d53-a2c0-d0562bf16a7d" />
intellisense should be suggesting `start_banks` and `start_track`.

**Comment by andrienndjibu-dev** [comment]
comment transformer une application dejà dévéloppé en application APK mobile ?
